drivers: gpio: nrfx: replace custom drive flags

gmarull opened the pull request:
Now that we have 8 bits reserved for vendor specific GPIO flags, it
doesn't make sense to keep the old drive "standard" nomenclature.
Instead, use a naming scheme the fits better with vendor hardware
capabilities.

This patch introduces a new set of flags to configure nRF GPIO drive
mode. The table below shows the equivalence between old and new flag
names:

| Old flags                                      | New flags             |
|------------------------------------------------|-----------------------|
| `NRF_GPIO_DS_DFLT_LOW`                         | `NRF_GPIO_DRIVE_S0`   |
| `NRF_GPIO_DS_DFLT_HIGH`                        | `NRF_GPIO_DRIVE_S1`   |
| `NRF_GPIO_DS_ALT_LOW`                          | `NRF_GPIO_DRIVE_H0`   |
| `NRF_GPIO_DS_ALT_HIGH`                         | `NRF_GPIO_DRIVE_H1`   |
| `NRF_GPIO_DS_DFLT`                             | `NRF_GPIO_DRIVE_S0S1` |
| `NRF_GPIO_DS_ALT`                              | `NRF_GPIO_DRIVE_H0H1` |
| `NRF_GPIO_DS_DFLT_LOW \| NRF_GPIO_DS_ALT_HIGH` | `NRF_GPIO_DRIVE_S0H1` |
| `NRF_GPIO_DS_ALT_LOW \| NRF_GPIO_DS_DFLT_HIGH` | `NRF_GPIO_DRIVE_H0S1` |

Documentation has also be re-written to explain in more detail the
meaning of the flags and how they can be used.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
drivers: gpio: nrfx: add custom drive flags

Now that we have 8 bits reserved for vendor specific GPIO flags,
introduce a new set of flags for nRF platforms to configure pins drive
mode. These new flags are equivalent to the previous existing ones, but
use a naming scheme the fits better with vendor hardware capabilities.

The table below shows the equivalence between old and new flag

| Old flags                 | New flags             |
|---------------------------|-----------------------|
| `NRF_GPIO_DS_DFLT_LOW`    | `NRF_GPIO_DRIVE_S0`   |
| `NRF_GPIO_DS_DFLT_HIGH`   | `NRF_GPIO_DRIVE_S1`   |
| `NRF_GPIO_DS_ALT_LOW`     | `NRF_GPIO_DRIVE_H0`   |
| `NRF_GPIO_DS_ALT_HIGH`    | `NRF_GPIO_DRIVE_H1`   |
| `NRF_GPIO_DS_DFLT`        | `NRF_GPIO_DRIVE_S0S1` |
| `NRF_GPIO_DS_ALT`         | `NRF_GPIO_DRIVE_H0H1` |
| `NRF_GPIO_DS_DFLT_LOW \|` | `NRF_GPIO_DRIVE_S0H1` |
| `NRF_GPIO_DS_ALT_HIGH`    |                       |
| `NRF_GPIO_DS_ALT_LOW \|`  | `NRF_GPIO_DRIVE_H0S1` |
| `NRF_GPIO_DS_DFLT_HIGH`   |                       |

Documentation has been written to explain in more detail the meaning of
the flags and how they can be used.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

henrikbrixandersen commented on include/zephyr/dt-bindings/gpio/nordic-nrf-gpio.h:
```diff
@@ -1,74 +1,64 @@
 /*
- * Copyright (c) 2022 Vestas Wind Systems A/S
+ * Copyright (c) 2022 Nordic Semiconductor ASA
```
Please do not remove copyright statements. Just add an additional line.

gmarull replied:
file has been re-written from scratch, there's no re-used content

henrikbrixandersen replied:
I think this is a slippery slope. If we go down this road we can start discussing what constitutes a total rewrite in all PRs and whether original copyrights should be kept or not. 

Say you changed half the file in one PR and the other half in a second PR. Should the second PR then remove all copyrights than yours?

Please leave existing copyrights in place.

carlocaione replied:
I agree with @gmarull here. The rationale I use in general is:
- If the PR is rewriting the file from scratch then the copyright can be changed
- If the PR is changing significantly the content of the file the copyright can be added on top
- If the PR is not changing significantly the content of the file then the copyright is not touched

I had this same discussion with someone else a few days ago :)

gmarull replied:
This file was created copying content from a file that did not have the `Copyright (c) 2022 Vestas Wind Systems A/S` copyright, so in theory it should have kept them as well. However now it's getting entirely re-written, so I think it's 100% Nordic owned content after the change. I understand what is copyrighted is the content.

stephanosio replied:
> file has been re-written from scratch, there's no re-used content

If that is the case, there should be two commits: one removing the existing file completely, and another one adding the new file.

carlocaione replied:
> If that is the case, there should be two commits: one removing the existing file completely, and another one adding the new file.

Please, only if that doesn't break bisectability.

gmarull replied:
> > file has been re-written from scratch, there's no re-used content
> 
> If that is the case, there should be two commits: one removing the existing file completely, and another one adding the new file.

done, I have done the change in 2 steps, first delete old flags and add _new_ ones.

henrikbrixandersen commented on include/zephyr/dt-bindings/gpio/nordic-nrf-gpio.h:
```diff
@@ -1,74 +1,64 @@
  * @{
  */
-/** @cond INTERNAL_HIDDEN */
-#define NRF_GPIO_DS_LOW_POS 9
-#define NRF_GPIO_DS_LOW_MASK (0x1U << NRF_GPIO_DS_LOW_POS)
```
Shouldn't all the existing flags be deprecated?

gmarull replied:
the generic ones are already deprecated, these are too recent (and not much used) to bother with deprecation IMO.

henrikbrixandersen replied:
The SoC specific flags are public API and they were introduced in v3.1.0. Users just had to go through changing their code once, with proper deprecation warnings. Why would we skip it now?

gmarull replied:
They were part of a stable API before, so it made sense to deprecate. The newly created ones were a 1:1 replacement  with vendor prefix, but not necessarily what vendors wanted now that it is possible to create custom flags. Unfortunately, we did not have time to adjust them, and since it's a feature rarely used (there are no uses in zephyr/sdk-nrf), deprecation of an _intermediate_ state doesn't sound like it is necessary.

gmarull committed:
drivers: gpio: nrfx: remove custom DS flags

The naming of the custom DS flags is not clear for nRF platforms. This
patch removes all the NRF_GPIO_DS* flags. New flags will be
re-introduced in a follow-up commit.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>
